### PR TITLE
feat(upload background image): Upload virtual background image.

### PR DESCRIPTION
**Pull request by tudor-phd**

### CODE REVIEW

File: react/features/virtual-background/components/VirtualBackgroundDialog.js [hunk 1]
```diff
@@ -54,23 +55,73 @@ type Props = {
...
+    };
+
+    const deleteStoredImage = image => {
+        setStoredImages(storedImages.filter(item => item !== image));
+    };
+
+    const uploadImage = async imageFile => {
+        const reader = new FileReader();
+
+        reader.readAsDataURL(imageFile[0]);
```
Comment (saghul) [inline react/features/virtual-background/components/VirtualBackgroundDialog.js:102]: Does this mean it will only upload the first image if I select several?

File: react/features/virtual-background/components/VirtualBackgroundDialog.js [hunk 2]
```diff
@@ -54,23 +55,73 @@ type Props = {
...
+                    tooltip: imageFile[0].name,
+                    name: imageFile[0].name,
+                    id: Math.random()
+                        .toString(36)
+                        .substring(2, 7),
+                    src: reader.result.toString()
+                }
+            ]);
+
+            await dispatch(setVirtualBackground(reader.result.toString(), true));
```
Comment (saghul) [inline react/features/virtual-background/components/VirtualBackgroundDialog.js:117]: We should do some size checks and save a resized version, we don't need an 8K image even if the user uploads one.

**Review by saghul — COMMENTED** [review]
Left some comments, PTAL!

File: react/features/virtual-background/components/VirtualBackgroundDialog.js (hunk 1 above)
Comment (tudor-phd) [reply react/features/virtual-background/components/VirtualBackgroundDialog.js:102]: You can't select multiple images for upload. The file input is set by default to accept a single file. The multiple property is not set.  Related to '[0]' -> this is how the file structure appears by default when you have a single file for upload.

File: react/features/virtual-background/components/VirtualBackgroundDialog.js [hunk 3]
```diff
@@ -54,23 +55,73 @@ type Props = {
...
+            isloading(true);
+            setStoredImages([
+                ...storedImages,
+                {
+                    tooltip: imageFile[0].name,
+                    name: imageFile[0].name,
+                    id: Math.random()
+                        .toString(36)
+                        .substring(2, 7),
+                    src: reader.result.toString()
```
[the remark this answers is not in the document]
Comment (tudor-phd) [reply react/features/virtual-background/components/VirtualBackgroundDialog.js:113]: I put toString() function because the virtualSource from action file is set to be a string and the base64(from uploaded image) is ArrayBuffer. I can remove this toString() function and let the type of virtualSource be 'any'

File: react/features/virtual-background/components/VirtualBackgroundDialog.js [hunk 4]
```diff
@@ -54,23 +55,73 @@ type Props = {
...
+    const localImages = localStorage.getItem('storedImages');
+    const [ storedImages, setStoredImages ] = useState((localImages && JSON.parse(localImages)) || []);
+    const [ loading, isloading ] = useState(false);
+
+
+    /**
+     * Updates stored images on local storage.
+     */
+    useEffect(() => {
+        localStorage.setItem('storedImages', JSON.stringify(storedImages));
```
[the remark this answers is not in the document]
Comment (tudor-phd) [reply react/features/virtual-background/components/VirtualBackgroundDialog.js:67]: I chose to use localStorage to persist images for future reuse. If the user closes the browser and opens it again he will have the images available and will not be forced to upload them again. Of course, if the user will clear the site data(clear cache) all images will disappear from local memory(local storage)

[the remark this answers is not in the document]
Comment (saghul) [reply react/features/virtual-background/components/VirtualBackgroundDialog.js:67]: We persist the redux store to local storage, so there is no need to do your own local storage :-)

File: react/features/virtual-background/components/VirtualBackgroundDialog.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (saghul) [reply react/features/virtual-background/components/VirtualBackgroundDialog.js:113]: We could use either, but I see we use it in 2 places, ideally we only do this conversion once.

File: react/features/stream-effects/virtual-background/JitsiStreamBackgroundEffect.js [hunk 5]
```diff
@@ -113,7 +113,14 @@ export default class JitsiStreamBackgroundEffect {
 
         this._outputCanvasCtx.globalCompositeOperation = 'destination-over';
         if (this._options.virtualBackground.isVirtualBackground) {
-            this._outputCanvasCtx.drawImage(this._virtualImage, 0, 0);
+            this._virtualImage.style.objectFit = 'cover';
```
Comment (saghul) [inline react/features/stream-effects/virtual-background/JitsiStreamBackgroundEffect.js:116]: As discussed, this won't work and we'll work on a proper "object-fit cover" rendering of the background image in the canvas.

File: react/features/virtual-background/actions.js [hunk 6]
```diff
@@ -41,11 +41,11 @@ export function toggleBackgroundEffect(enabled: boolean) {
  * @param {boolean} isVirtualBackground - Indicate if virtual image is activated.
  * @returns {{
  *     type: SET_VIRTUAL_BACKGROUND,
- *     virtualSource: string,
+ *     virtualSource: any,
```
Comment (saghul) [inline react/features/virtual-background/actions.js:44]: Why the change, the base64 encoded data is a string too.

File: react/features/virtual-background/components/VirtualBackgroundDialog.js [hunk 7]
```diff
@@ -54,23 +59,97 @@ type Props = {
  * @returns {ReactElement}
  */
 function VirtualBackground({ dispatch, t }: Props) {
+    const localImages = jitsiLocalStorage.getItem('virtualBackgrounds');
+    const [ storedImages, setStoredImages ] = useState((localImages && JSON.parse(localImages)) || []);
+    const [ loading, isloading ] = useState(false);
+
+    const toDataURL = url =>
+        fetch(url)
```
Comment (saghul) [inline react/features/virtual-background/components/VirtualBackgroundDialog.js:67]: can you please write this as an async function and use await instead? Also, take it out of the component, it's generic and can go in functions.js

File: react/features/virtual-background/components/VirtualBackgroundDialog.js [hunk 8]
```diff
@@ -54,23 +59,97 @@ type Props = {
...
+    const deleteStoredImage = image => {
+        setStoredImages(storedImages.filter(item => item !== image));
+    };
+
+    /**
+     * Updates stored images on local storage.
+     */
+    useEffect(() => {
+        jitsiLocalStorage.setItem('virtualBackgrounds', JSON.stringify(storedImages));
+        if (storedImages.length === 22) {
```
Comment (saghul) [inline react/features/virtual-background/components/VirtualBackgroundDialog.js:89]: Put this constant at the top of the file and leave a comment about why it's 22 please.

File: react/features/virtual-background/components/VirtualBackgroundDialog.js [hunk 9]
```diff
@@ -54,23 +59,97 @@ type Props = {
...
+        isloading(true);
         setSelected('none');
-        dispatch(setVirtualBackground('', false));
-        dispatch(toggleBackgroundEffect(false));
+        await dispatch(setVirtualBackground(null, false));
+        await dispatch(toggleBackgroundEffect(false));
+        isloading(false);
+    };
+
+    const addImageBackground = async image => {
```
Comment (saghul) [inline react/features/virtual-background/components/VirtualBackgroundDialog.js:111]: `setUploadedImageBackground`

File: react/features/virtual-background/components/VirtualBackgroundDialog.js [hunk 10]
```diff
@@ -54,23 +59,97 @@ type Props = {
...
+        isloading(true);
+        setSelected(image.id);
+        await dispatch(setVirtualBackground(image.src, true));
+        await dispatch(toggleBackgroundEffect(true));
+        isloading(false);
     };
 
-    const addImageBackground = image => {
+    const addConvertedImageBackground = async image => {
```
Comment (saghul) [inline react/features/virtual-background/components/VirtualBackgroundDialog.js:119]: `setImageBackground`

File: react/features/virtual-background/components/VirtualBackgroundDialog.js [hunk 11]
```diff
@@ -54,23 +59,97 @@ type Props = {
...
+
+        reader.readAsDataURL(imageFile[0]);
+        reader.onload = async () => {
+            const resizedImage = await resizeImage(reader.result);
+
+            isloading(true);
+            setStoredImages([
+                ...storedImages,
+                {
+                    tooltip: imageFile[0].name,
```
Comment (saghul) [inline react/features/virtual-background/components/VirtualBackgroundDialog.js:138]: No tooltips for the uploaded images.

File: react/features/virtual-background/components/VirtualBackgroundDialog.js [hunk 12]
```diff
@@ -54,23 +59,97 @@ type Props = {
...
+                    name: imageFile[0].name,
```
Comment (saghul) [inline react/features/virtual-background/components/VirtualBackgroundDialog.js:139]: Why do we need the name btw?

File: react/features/virtual-background/components/VirtualBackgroundDialog.js [hunk 13]
```diff
@@ -79,38 +158,84 @@ function VirtualBackground({ dispatch, t }: Props) {
...
-                    content = { t('virtualBackground.removeBackground') }
-                    position = { 'top' }>
-                    <div
-                        className = { selected === 'none' ? 'none-selected' : 'virtual-background-none' }
-                        onClick = { () => removeBackground() }>
-                        None
+            {loading ? (
+                <div>
+                    <span>{t('virtualBackground.pleaseWait')}</span>
+                    <Spinner
```
Comment (saghul) [inline react/features/virtual-background/components/VirtualBackgroundDialog.js:164]: Leave some space between the text and the spinner.

File: react/features/virtual-background/components/VirtualBackgroundDialog.js [hunk 14]
```diff
@@ -79,38 +158,84 @@ function VirtualBackground({ dispatch, t }: Props) {
...
+            ) : (
+                <div>
+                    <div className = 'virtual-background-dialog'>
+                        <Tooltip
+                            content = { t('virtualBackground.removeBackground') }
+                            position = { 'top' }>
+                            <div
+                                className = { selected === 'none' ? 'none-selected' : 'virtual-background-none' }
+                                onClick = { removeBackground }>
+                                None
```
Comment (saghul) [inline react/features/virtual-background/components/VirtualBackgroundDialog.js:177]: Apply a translation to this. Or maybe we can use an icon to not translate it at all.

File: react/features/virtual-background/components/VirtualBackgroundDialog.js [hunk 15]
```diff
@@ -79,38 +158,84 @@ function VirtualBackground({ dispatch, t }: Props) {
...
+                            content = { t('virtualBackground.enableBlur') }
+                            position = { 'top' }>
+                            <Icon
+                                className = { selected === 'blur' ? 'blur-selected' : '' }
+                                onClick = { () => enableBlur() }
+                                size = { 50 }
+                                src = { IconBlurBackground } />
+                        </Tooltip>
+                        {images.map((image, index) => (
+                            <Tooltip
```
Comment (saghul) [inline react/features/virtual-background/components/VirtualBackgroundDialog.js:190]: No tooltip for uploaded images.

File: react/features/virtual-background/components/VirtualBackgroundDialog.js [hunk 16]
```diff
@@ -79,38 +158,84 @@ function VirtualBackground({ dispatch, t }: Props) {
...
+                                <img
+                                    className = { selected === image.id ? 'thumbnail-selected' : 'thumbnail' }
+                                    onClick = { () => addImageBackground(image) }
+                                    onError = { event => event.target.style.display = 'none' }
+                                    src = { image.src } />
+                                <Icon
+                                    className = { 'delete-image-icon' }
+                                    onClick = { () => deleteStoredImage(image) }
+                                    size = { 10 }
+                                    src = { IconTrash } />
```
Comment (saghul) [inline react/features/virtual-background/components/VirtualBackgroundDialog.js:233]: Lets put the icon in the top right, on hover.

File: react/features/virtual-background/functions.js [hunk 17]
```diff
@@ -20,3 +18,95 @@ export function checkBlurSupport() {
...
+
+    const halfImageDimensions = {
+        width: 0,
+        height: 0
+    };
+
+    // Quickly reduce the dize by 50% each time in few iterations until the size is less then
+    // 2x time the target size - the motivation for it, is to reduce the aliasing that would have been
+    // created with direct reduction of very big image to small image.
+    while (curImageDimensions.width * 0.5 > width) {
```
Comment (saghul) [inline react/features/virtual-background/functions.js:70]: As discussed, remove this. This function should just make sure the image fits in the given canvas size, compress it and base64 encode it.

File: react/features/virtual-background/functions.js [hunk 18]
```diff
@@ -20,3 +18,95 @@ export function checkBlurSupport() {
...
+        0,
+        curImageDimensions.width,
+        curImageDimensions.height,
+        0,
+        0,
+        width,
+        height
+    );
+
+    return outputCanvas.toDataURL('image/jpeg', 0.1);
```
Comment (saghul) [inline react/features/virtual-background/functions.js:111]: Isn't 0.1 a bit low? Maybe let's leave the default?

**Review by saghul — CHANGES_REQUESTED** [review]
Left some comments, PTAL.

File: react/features/virtual-background/functions.js (hunk 18 above)
Comment (tudor-phd) [reply react/features/virtual-background/functions.js]: Changed to medium(0.5) This will save space on local storage.

File: react/features/virtual-background/functions.js (hunk 17 above)
Comment (tudor-phd) [reply react/features/virtual-background/functions.js]: Simplified the entire function.

File: react/features/virtual-background/components/VirtualBackgroundDialog.js (hunk 15 above)
Comment (tudor-phd) [reply react/features/virtual-background/components/VirtualBackgroundDialog.js]: Removed tooltips from all thumbnails

File: react/features/virtual-background/components/VirtualBackgroundDialog.js (hunk 12 above)
Comment (tudor-phd) [reply react/features/virtual-background/components/VirtualBackgroundDialog.js]: Removed the name. Also the tooltip.

**Review by saghul — APPROVED** [review]
